File: android/app/src/main/kotlin/com/shibu/app/data/Kanji.kt
package com.shibu.app.data

import org.json.JSONObject

/**
 * One entry from the bundled kanji deck.
 *
 * Field names are deliberately short because the same JSON is parsed on the
 * Dart side and shipped inside the APK; see assets/data/kanji.json.
 */
data class Kanji(
    val id: Int,
    /** The kanji character itself, e.g. 駅. */
    val character: String,
    /** JLPT level: 5 is easiest, 3 is hardest in the bundled deck. */
    val level: Int,
    val strokes: Int,
    /** Primary reading in romaji, e.g. "eki". */
    val romaji: String,
    /** Primary reading in kana, e.g. えき. */
    val kana: String,
    /** English meaning, e.g. "station". */
    val meaning: String,
    val onyomi: String,
    val kunyomi: String,
    /** Example compound, e.g. 駅前. */
    val example: String,
    val exampleKana: String,
    val exampleRomaji: String,
    val exampleMeaning: String,
) {
    /** The reading line as shown on the widget: "eki・えき". */
    val readingLine: String get() = "$romaji・$kana"

    /** The example line as shown on the widget: "駅前・in front of station". */
    val exampleLine: String get() = "$example・$exampleMeaning"

    companion object {
        fun fromJson(o: JSONObject): Kanji = Kanji(
            id = o.getInt("id"),
            character = o.getString("c"),
            level = o.getInt("l"),
            strokes = o.optInt("s", 0),
            romaji = o.getString("r"),
            kana = o.getString("k"),
            meaning = o.getString("m"),
            onyomi = o.optString("on", ""),
            kunyomi = o.optString("kun", ""),
            example = o.getString("ex"),
            exampleKana = o.optString("exk", ""),
            exampleRomaji = o.optString("exr", ""),
            exampleMeaning = o.getString("exm"),
        )
    }
}

File: lib/models/kanji.dart
/// One entry from the bundled kanji deck.
///
/// The JSON uses short keys because the same file is parsed by the native
/// wallpaper and widget code; see `assets/data/kanji.json`.
class Kanji {
  const Kanji({
    required this.id,
    required this.character,
    required this.level,
    required this.strokes,
    required this.romaji,
    required this.kana,
    required this.meaning,
    required this.onyomi,
    required this.kunyomi,
    required this.example,
    required this.exampleKana,
    required this.exampleRomaji,
    required this.exampleMeaning,
  });

  final int id;

  /// The character itself, e.g. 駅.
  final String character;

  /// JLPT level: 5 is the easiest, 3 the hardest in the bundled deck.
  final int level;
  final int strokes;

  /// Primary reading in romaji, e.g. `eki`.
  final String romaji;

  /// Primary reading in kana, e.g. えき.
  final String kana;
  final String meaning;
  final String onyomi;
  final String kunyomi;

  /// Example compound, e.g. 駅前.
  final String example;
  final String exampleKana;
  final String exampleRomaji;
  final String exampleMeaning;

  /// The reading line as shown on the card: `eki・えき`.
  String get readingLine => '$romaji・$kana';

  /// The example line as shown on the card: `駅前・in front of station`.
  String get exampleLine => '$example・$exampleMeaning';

  /// Everything a search should match against.
  String get searchIndex => [
    character,
    romaji,
    kana,
    meaning,
    onyomi,
    kunyomi,
    example,
    exampleRomaji,
    exampleMeaning,
  ].join(' ').toLowerCase();

  factory Kanji.fromJson(Map<String, dynamic> json) => Kanji(
    id: json['id'] as int,
    character: json['c'] as String,
    level: json['l'] as int,
    strokes: (json['s'] as num?)?.toInt() ?? 0,
    romaji: json['r'] as String,
    kana: json['k'] as String,
    meaning: json['m'] as String,
    onyomi: json['on'] as String? ?? '',
    kunyomi: json['kun'] as String? ?? '',
    example: json['ex'] as String,
    exampleKana: json['exk'] as String? ?? '',
    exampleRomaji: json['exr'] as String? ?? '',
    exampleMeaning: json['exm'] as String,
  );
}
